integration/aggregate_query_compat_test.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"testing"
+
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+func TestAggregateCompatMatchExpr(t *testing.T) {
```

Comment:
@AlekSi do you have a particular example? I'm afraid I have to disagree with your current statement, at least because the second part `db.col.aggregate(X)` should include an aggregation pipeline stage, and `X` is not an aggregation pipeline stage but an aggregation expression. Your first statement is not entirely true either because `aggregate` needs an array of stages, not an object.

Regarding the `$match` stage, see also: https://www.mongodb.com/docs/manual/reference/operator/aggregation/match/#restrictions.

Reply:
Do you mean `db.col.aggregate([{$match: {$expr: X}}])` and `db.col.find({$expr: X})` are the same? Yes they are the same as far as those tests show, I didn't know it until I added those tests.

Reply:
All that goes back to my message in Slack. The fact that we try to get an understanding from examples rather than from fundamentals is even more concerning.

`$match` is the aggregation stage that uses query operators for filtering. `$expr` is a query operator that uses aggregation expressions. They mirror each other pretty much by definition. That test pretty much tests that `(√x)² == x` for 20 normal values of x. I'm not saying that we should not test it for _one_ normal x and some edge cases. But we already have a problem with slow tests that no amount of engineering could solve if we keep adding tautologous tests.

Reply:
Yes I am aware they mirror the behavior and that tests do test them both. Okie, I'm happy with reducing the number of tests by only testing few scenarios of the other one.

Reply:
I totally agree with Chi. We don't know if things are equal unless we test them. Especially, in case of error codes, error messages, etc. We have already had examples in the past when we thought things work in a way A (because documentation was silent about it), but it turned out in reality they work in a way B. I understand a concern regarding the number of tests, if we are ready for potential incompatibility, let's reduce the tests, it's never a problem to delete something.

Regarding "who uses `$match` in aggregation when it's possible to use `find`" - well, I guess someone who needs more than one stage in the aggregation pipeline. E.g. if after `$match` I want to group, or project, or something, I would prefer to have `$match` with `$expr`, not `find` and manual processing of data.

Reply:
For the case of `$match` stage and `find` command, it's safe to remove many of those tests, we've validated that they are the same. So let's remove those duplicated ones and not hinder the performance :hugs: 

During the implementation though, those tests were still helpful and it doesn't hurt to test them. But we don't need them anymore :hugs:

Reply:
> We don't know if things are equal unless we test them

The fact that MongoDB has some unexpected behavior in some edge cases does not mean that everything is an edge case. We don't test 2+2 and then test 2+3 to check if they behave differently; instead, we test 2+100500 because we suspect the overflow. So why do we test `true` and then test `false` without testing `null`, for example?

> it's never a problem to delete something

That requires much more thinking and analysis after the fact – is it safe to delete some tests? does it test some edge cases? why this test exists in the first place? It is not obvious that the test exists for no good reason. That's why we should not add them in the first place.

> Regarding "who uses $match in aggregation when it's possible to use find"

I did not say that, though.